cli/lsp/registries.rs

```diff
@@ -654,339 +612,366 @@ impl ModuleRegistry {
   /// any, for the specifier.
   pub async fn get_completions(
     &self,
-    current_specifier: &str,
-    offset: usize,
+    text: &str,
     range: &lsp::Range,
+    resolved: Option<&ModuleSpecifier>,
     specifier_exists: impl Fn(&ModuleSpecifier) -> bool,
   ) -> Option<lsp::CompletionList> {
-    if let Ok(specifier) = Url::parse(current_specifier) {
-      let origin = base_url(&specifier);
-      let origin_len = origin.chars().count();
-      if offset >= origin_len {
-        if let Some(registries) = self.origins.get(&origin) {
-          let path = &specifier[Position::BeforePath..];
-          let path_offset = offset - origin_len;
-          let mut completions = HashMap::<String, lsp::CompletionItem>::new();
-          let mut is_incomplete = false;
-          let mut did_match = false;
-          for registry in registries {
-            let tokens = parse(&registry.schema, None)
-              .map_err(|e| {
-                error!(
-                  "Error parsing registry schema for origin \"{}\". {}",
-                  origin, e
-                );
-              })
-              .ok()?;
-            let mut i = tokens.len();
-            let last_key_name = StringOrNumber::String(
-              tokens
-                .iter()
-                .last()
-                .map(|t| {
-                  if let Token::Key(key) = t {
-                    if let StringOrNumber::String(s) = &key.name {
-                      return s.clone();
-                    }
-                  }
-                  "".to_string()
-                })
-                .unwrap_or_default(),
+    let resolved = resolved
+      .map(Cow::Borrowed)
+      .or_else(|| ModuleSpecifier::parse(text).ok().map(Cow::Owned))?;
+    let resolved_str = resolved.as_str();
+    let origin = base_url(&resolved);
+    let origin_len = origin.chars().count();
+    let registries = self.origins.get(&origin)?;
+    let path = &resolved[Position::BeforePath..];
+    let path_offset = resolved_str.chars().count() - origin_len;
```

Comment:
It might be better to use names like `path_char_offset` and `origin_char_count`. Usually `len` means byte len.